<div class="alert alert-block alert-info">
<h3>Student Information</h3> Please provide information about yourself.<br>
<b>Name</b>: Avantika Yellapantula<br>
<b>NetID</b>: ay257<br>
<b>Recitation (1 or 2)</b>: 1<br>
<b>Notes to Grader</b> (optional):<br>
<br><br>
<b>IMPORTANT</b>
Your work will not be graded withour your initials below<br>
I certify that this lab represents my own work and I have read the RU academic intergrity policies at<br>
<a href="https://www.cs.rutgers.edu/academic-integrity/introduction">https://www.cs.rutgers.edu/academic-integrity/introduction </a><br>
<b>Initials</b>: AY


<h3>Grader Notes</h3>
<b>Your Grade<b>:<br>
<b>Grader Initials</b>:<br>
<b>Grader Comments</b> (optional):<br>
</div>

## CS 439 - Introduction to Data Science
### Spring 2019

# Lab 2: Working with Pandas

### Due Date : Thursday February 7, 2019 by 11:59 PM.###

### Instructions


This is an introductory lab on Python/Pandas. The purpose of this lab is to learn how to use 
some basic commands in Pandas in your data science projects. We will also do some data analysis with a real dataset of babynames. This lab is presented as a Jupyter notebook. 
Please execute the cells (SHIFT-ENTER) that are already completed and then your task is to fill in the code
between ### BEGIN SOLUTION ### and ### END SOLUTION ###


[Pandas](https://pandas.pydata.org/) is one of the most widely used Python libraries in data science. In this lab, you will learn commonly used data wrangling operations/tools in Pandas. We aim to give you familiarity with:

* Creating dataframes
* Slicing data frames (ie. selecting rows and columns)
* Filtering data (using boolean arrays)
* Data Aggregation/Grouping dataframes
* Merging dataframes

In this lab, you are going to use several pandas methods like `drop()`, `loc[]`, `groupby()`. You may press `shift+tab` on the method parameters to see the documentation for that method.

**Just as a side note**: Pandas operations can be confusing at times and the documentation is not that great, but it is OK to be stumped when figuring out why a piece of code is not doing what it's supposed to. We don't expect you to memorize all the different Pandas functions, just know the basic ones like `iloc[]`, `loc[]`, slicing, and other general dataframe operations. For exams and quizzes, you will get a Pandas cheat sheet.

Throughout the semester, you will have to search through Pandas documentation and experiment, but remember it is part of the learning experience and will help shape you as a data scientist!  You need to learn how to read documentation.

## Setup
The first step in this lab is to setup the environment. We will read in pandas, numpy and matplotlib

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Creating DataFrames & Basic Manipulations

A [dataframe](http://pandas.pydata.org/pandas-docs/stable/dsintro.html#dataframe) is a two-dimensional labeled data structure with columns of potentially different types.

The pandas [`DataFrame` function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html) provides at least two syntaxes to create a data frame.

**Syntax 1: ** You can create a data frame by specifying the columns and values using a dictionary as shown below. 

The keys of the dictionary are the column names, and the values of the dictionary are lists containing the row entries.

In [2]:
fruit_info = pd.DataFrame(
    data={'fruit': ['apple', 'orange', 'banana', 'raspberry'],
          'color': ['red', 'orange', 'yellow', 'pink']
          })
fruit_info

fruit   color
0      apple     red
1     orange  orange
2     banana  yellow
3  raspberry    pink

**Syntax 2: ** You can also define a dataframe by specifying the rows like below. 

Each row corresponds to a distinct tuple, and the columns are specified separately.

In [3]:
fruit_info2 = pd.DataFrame(
    [("red", "apple"), ("orange", "orange"), ("yellow", "banana"),
     ("pink", "raspberry")], 
    columns = ["color", "fruit"])
fruit_info2

color      fruit
0     red      apple
1  orange     orange
2  yellow     banana
3    pink  raspberry

You can obtain the dimensions of a data frame by using the shape attribute dataframe.shape

In [4]:
(num_rows, num_columns) = fruit_info.shape
num_rows, num_columns

(4, 2)

# Task 1 - Panda Basics

### Activity 1.1 - adding columns

You can add a column by `dataframe['new column name'] = [data]`. Please add a column called `rank1` to the `fruit_info` table which contains a 1,2,3, or 4 based on your personal preference ordering for each fruit. 


In [6]:
### BEGIN SOLUTION
fruit_info['rank1']= [2,1,3,4]
### END SOLUTION

In [7]:
# print the data frame to see if the column has been added
fruit_info

fruit   color  rank1
0      apple     red      2
1     orange  orange      1
2     banana  yellow      3
3  raspberry    pink      4

In [12]:
### BEGIN SOLUTION
# complete the assert statements below
assert fruit_info["rank1"].dtype == 'int64'
assert len(fruit_info["rank1"].dropna()) !=0 
### END SOLUTION

### Activity 1.2 - adding columns II

You can ALSO add a column by `dataframe.loc[:, 'new column name'] = [data]`. This way to modify an existing dataframe is preferred over the assignment above. In other words, it is best that you use `loc[]`. Although using `loc[]` is more verbose, it is faster. (However, this tradeoff is more likely to be valuable in production than during interactive use.) In lecture we will explain in more detail what `loc[]` does, but essentially, the first parameter is for the rows and second is for columns. The `:` means keep all rows and the `new column name` indicates the column you are modifying or in this case adding. 

Please add a column called `rank2` to the `fruit_info` table which contains a 1,2,3, or 4 based on your personal preference ordering for each fruit. Use the loc to do this.


In [13]:
### BEGIN SOLUTION
fruit_info.loc[:,'rank2']=[3,1,2,4]
### END SOLUTION

In [14]:
fruit_info

fruit   color  rank1  rank2
0      apple     red      2      3
1     orange  orange      1      1
2     banana  yellow      3      2
3  raspberry    pink      4      4

In [15]:
### BEGIN SOLUTION
# complete the assert statements below
assert fruit_info["rank2"].dtype == 'int64'
assert len(fruit_info["rank2"].dropna()) != 0
### END SOLUTION

### Activity 1.3 - dropping columns

Use the `.drop()` method to [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html) the both the `rank1` and `rank2` columns you created. (Make sure to use the `axis` parameter correctly) 

Hint: Look through the documentation to see how you can drop multiple columns of a Pandas dataframe at once, it may involve a list.

In [16]:
### BEGIN SOLUTION
fruit_info_original=fruit_info.drop(['rank1','rank2'],axis=1)
### END SOLUTION

In [17]:
fruit_info_original

fruit   color
0      apple     red
1     orange  orange
2     banana  yellow
3  raspberry    pink

In [19]:
### BEGIN SOLUTION
# complete the assert statements below
assert fruit_info_original.shape[1] !=0
### END SOLUTION

### Activity 1.4 - Renaming columns

Use the `.rename()` method to [rename](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html) the columns of `fruit_info_original` so they begin with a capital letter. Set the `inplace` parameter correctly to change the `fruit_info_original` dataframe. (**Hint:** in Question 2, `drop` creates and returns a new dataframe instead of changing `fruit_info` because `inplace` by default is `False`)

In [25]:
### BEGIN SOLUTION
fruit_info_original.rename(index=str, columns={"fruit": "Fruit", "color": "Color"}, inplace='true')
### END SOLUTION

In [26]:
fruit_info_original

Fruit   Color
0      apple     red
1     orange  orange
2     banana  yellow
3  raspberry    pink

In [27]:
### BEGIN SOLUTION
# complete the assert statements below
assert fruit_info_original.columns[0] != 0 
assert fruit_info_original.columns[1] != 0
### END SOLUTION

# Task 2 - Babyname case study
Now that we have learned the basics of data frames, let's move on to a babynames dataset. The babynames data set contains babynames from all states. We will clean and wrangle a data frame with babynames data for the remainder of this lab.

## Helper Function
### `fetch_and_cache` Helper

Downloading large files take time. If you have to download the file everytime you execute this notebook, things can be slow and inefficient. To do this task once, we will use the following function for downloading and caching the data in the `data/` directory and returns the `Path` to the downloaded file. We can use this function throughout the course, so that we save the download bandwidth. This function checks the local data directory for the file name, before attempting to download. Note that, going forward, we will maintain a "data" folder in our lab working folder. 

In [3]:
# Author: Berkeley Data Science Group
def fetch_and_cache(data_url, file, data_dir="data", force=False):
    """
    Download and cache a url and return the file object.
    
    data_url: the web address to download
    file: the file in which to save the results.
    data_dir: (default="data") the location to save the data
    force: if true the file is always re-downloaded 
    
    return: The pathlib.Path object representing the file.
    """
    import requests
    from pathlib import Path
    data_dir = Path(data_dir)
    data_dir.mkdir(exist_ok=True)
    file_path = data_dir/Path(file)
    if force and file_path.exists():
        file_path.unlink()
    if force or not file_path.exists():
        print('Downloading...', end=' ')
        resp = requests.get(data_url)
        with file_path.open('wb') as f:
            f.write(resp.content)
        print('Done!')
    else:
        import time 
        birth_time = time.ctime(file_path.stat().st_ctime)
        print("Using cached version downloaded:", birth_time)
    return file_path

Let us step through the above function and identify some of the key lines.

In Python, whenever you want to check if a file exists in a certain path, it is not sufficient to just have the string representation of the path, you need to create a Path object usign the `Path()` constructor. Essentially, after the Path object is created for the directory, a directory is created at that path location using the `mkdir()` method. Then, within the directory, a path for the file itself is created and if the path has already been linked (a.k.a file has already been created and put in the directory), then a new one is not created and instead uses the cached version.

The function `exists()` in the code above is one way to check if a file exists at a certain path when called on a path object. There is also another way this can be done using the `os` library in Python. If you decided to use the `os` library, you wouldn't need to create a Path object and rather pass in the the string representation of the path.

Now, going back to the code, if the path hasn't been linked, then the file is downloaded and created at the path location. 

The benefit of this function is that not only can you force when you want a new file to be downloaded using the `force` parameter, but in cases when you don't need the file to be re-downloaded, you can use the cached version and save download time.

Below we use fetch and cache to download the `namesbystate.zip` zip file. 

**This might take a little while!**

First let's run the following cells to build the dataframe `baby_names`.
The cells below download the data from the web and extract the data in a New Jersey region. There should be a total of ???  records.

## Activity 2.1-  Download Babynames Data Set
The zip file is approximately 21 MB. It might take a little time to download.But once you downloaded, it should always be available in the data folder for your lab.

In [4]:
data_url = 'https://www.ssa.gov/oact/babynames/state/namesbystate.zip'
# use fetct_and_cache to avoid multiple downloads.
namesbystate_path = fetch_and_cache(data_url, 'namesbystate.zip')

Using cached version downloaded: Thu Feb  7 10:28:33 2019


The following cell builds the final full `baby_names` DataFrame. Here is documentation for [pd.concat](https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.concat.html) if you want to know more about its functionality. 


In [10]:
import pandas as pd
import zipfile
#filename="namesbystates.zip"
zf = zipfile.ZipFile("data/namesbystate.zip", 'r')

field_names = ['State', 'Sex', 'Year', 'Name', 'Count']

def load_dataframe_from_zip(zf, f):
    with zf.open(f) as fh: 
        return pd.read_csv(fh, header=None, names=field_names)

# List comprehension
states = [
    load_dataframe_from_zip(zf, f)
    for f in sorted(zf.filelist, key=lambda x:x.filename) 
    if f.filename.endswith('.TXT')
]

baby_names = pd.concat(states).reset_index(drop=True)

In [11]:
# print key information about this data frame
baby_names.head()
len(baby_names)

5933561

In [8]:
# find the number of records in the file
assert(len(baby_names) == 5933561)

## Activity 2.2 - Slicing Data Frames - Selecting rows and columns


### Selection Using Label

**Column Selection** 
To select a column of a `DataFrame` by column label, the safest and fastest way is to use the `.loc` [method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.loc.html). General usage looks like `frame.loc[rowname,colname]`. (Reminder that the colon `:` means "everything").  For example, if we want the `color` column of the `ex` data frame, we would use : `ex.loc[:, 'color']`

- You can also slice across columns. For example, `baby_names.loc[:, 'Name':]` would give select the columns `Name` and the columns after.

- *Alternative:* While `.loc` is invaluable when writing production code, it may be a little too verbose for interactive use. One recommended alternative is the `[]` method, which takes on the form `frame['colname']`.

**Row Selection**
Similarly, if we want to select a row by its label, we can use the same `.loc` method. In this case, the "label" of each row refers to the index (ie. primary key) of the dataframe.

In [12]:
# Select rows 2:5 and the column 'Name'
### BEGIN SOLUTION
baby_names.loc[2:5, 'Name':]
### END SOLUTION


Name  Count
2      Anna     10
3  Margaret      8
4     Helen      7
5     Elsie      6

In [13]:
# Select rows 2:5 and column names using a list ['name']
### BEGIN SOLUTION
baby_names.loc[2:5,:]
### END SOLUTION


State Sex  Year      Name  Count
2    AK   F  1910      Anna     10
3    AK   F  1910  Margaret      8
4    AK   F  1910     Helen      7
5    AK   F  1910     Elsie      6

The `.loc` actually uses the Pandas row index rather than row id/position of rows in the dataframe to perform the selection. Also, notice that if you write `2:5` with `loc[]`, contrary to normal Python slicing functionality, the end index is included, so you get the row with index 5. 


There is another Pandas slicing function called `iloc[]` which lets you slice the dataframe by row id and column id instead of by column name and row index (for `loc[]`). This is really the main difference between the 2 functions and it is important that you remember the difference and why you might want to use one over the other. 

In addition, with `iloc[]`, the end index is NOT included, like with normal Python slicing.

Here is an example of how we would get the 2nd, 3rd, and 4th rows with only the `Name` column of the `baby_names` dataframe using both `iloc[]` and `loc[]`. Observe the difference.

In [14]:
# Select row indices 1:4 and column idex 3 using iloc
### BEGIN SOLUTION
baby_names.iloc[1:4,3]
### END SOLUTION


1       Annie
2        Anna
3    Margaret
Name: Name, dtype: object

In [15]:
# select row indices 1:3 and column "name" using loc
### BEGIN SOLUTION
baby_names.loc[1:3, 'Name':]
### END SOLUTION


Name  Count
1     Annie     12
2      Anna     10
3  Margaret      8

Lastly, we can change the index of a dataframe using the `set_index` method.

In [24]:
# Example: We change the index from 0,1,2... to the Name column
df = baby_names[:5].set_index("Name") 
df

State Sex  Year  Count
Name                           
Mary        AK   F  1910     14
Annie       AK   F  1910     12
Anna        AK   F  1910     10
Margaret    AK   F  1910      8
Helen       AK   F  1910      7

We can now lookup rows by name directly:

In [17]:
df.loc[['Mary', 'Anna'], :]

State Sex  Year  Count
Name                       
Mary    AK   F  1910     14
Anna    AK   F  1910     10

However, if we still want to access rows by location we will need to use the integer loc (`iloc`) accessor:

In [36]:
# Example: 
# df.loc[2:5,"Year"] You can't do this
df.iloc[1:4,2:3]

Year
Name          
Annie     1910
Anna      1910
Margaret  1910

## Activity 2.3
Selecting multiple columns is easy.  You just need to supply a list of column names.  Select the `Name` and `Year` **in that order** from the `baby_names` table.

In [44]:
### BEGIN SOLUTION
name_and_year=baby_names[:].set_index("Name")
name_and_year.iloc[:,2:3]
### END SOLUTION

Year
Name           
Mary       1910
Annie      1910
Anna       1910
Margaret   1910
Helen      1910
Elsie      1910
Lucy       1910
Dorothy    1910
Mary       1911
Margaret   1911
Ruth       1911
Annie      1911
Elizabeth  1911
Helen      1911
Mary       1912
Elsie      1912
Agnes      1912
Anna       1912
Helen      1912
Louise     1912
Jean       1912
Ruth       1912
Alice      1912
Esther     1912
Ethel      1912
Margaret   1912
Marie      1912
Mary       1913
Elizabeth  1913
Margaret   1913
...         ...
Max        2017
Maxwell    2017
Raiden     2017
Reed       2017
Ronan      2017
Ryder      2017
Sebastian  2017
Tucker     2017
Tyler      2017
Waylon     2017
Abel       2017
Austin     2017
Brayden    2017
Bryson     2017
Charlie    2017
Declan     2017
Ellis      2017
Gunnar     2017
Jayce      2017
Juan       2017
Judah      2017
Jude       2017
Kaison     2017
Kingston   2017
Lane       2017
Nathan     2017
Rayden     2017
Sterling   2017
Timothy    2017
Zander     2017

[5933561 rows x 1 columns]

In [45]:
name_and_year.shape

(5933561, 4)

In [48]:
### BEGIN SOLUTION
# complete the assert statements below
assert name_and_year.shape !=0
assert name_and_year.loc[0,"Name"] ==
assert name_and_year.loc[0,"Year"] ==
### END SOLUTION

SyntaxError: invalid syntax (<ipython-input-48-f37c87382f91>, line 4)

As you may have noticed above, the .loc() method is a way to re-order the columns within a dataframe.

## Activity 2.4 - Filtering Data

### Filtering with boolean arrays

Filtering is the process of removing unwanted material.  In your quest for cleaner data, you will undoubtedly filter your data at some point: whether it be for clearing up cases with missing values, culling out fishy outliers, or analyzing subgroups of your data set.  Note that compound expressions have to be grouped with parentheses. Example usage looks like `df[df[column name] < 5]]`.

For your reference, some commonly used comparison operators are given below.

Symbol | Usage      | Meaning 
------ | ---------- | -------------------------------------
==   | a == b   | Does a equal b?
<=   | a <= b   | Is a less than or equal to b?
>=   | a >= b   | Is a greater than or equal to b?
<    | a < b    | Is a less than b?
&#62;    | a &#62; b    | Is a greater than b?
~    | ~p       | Returns negation of p
&#124; | p &#124; q | p OR q
&    | p & q    | p AND q
^  | p ^ q | p XOR q (exclusive or)

In the following we construct the DataFrame containing only names registered in New Jersey

In [56]:
### BEGIN SOLUTION
# construct a DataFrame containing only the names registered in NJ. 
# The table below shows the expected answer.
nj=baby_names.loc[baby_names['State']=='NJ']
nj
### END SOLUTION

State Sex  Year       Name  Count
3501480    NJ   F  1910       Mary    593
3501481    NJ   F  1910      Helen    438
3501482    NJ   F  1910       Anna    355
3501483    NJ   F  1910   Margaret    311
3501484    NJ   F  1910  Elizabeth    260
3501485    NJ   F  1910    Dorothy    255
3501486    NJ   F  1910       Rose    201
3501487    NJ   F  1910       Ruth    188
3501488    NJ   F  1910    Mildred    174
3501489    NJ   F  1910   Florence    169
3501490    NJ   F  1910  Catherine    158
3501491    NJ   F  1910      Marie    152
3501492    NJ   F  1910    Lillian    130
3501493    NJ   F  1910      Alice    125
3501494    NJ   F  1910    Frances    124
3501495    NJ   F  1910  Josephine    103
3501496    NJ   F  1910      Ethel    101
3501497    NJ   F  1910     Marion    100
3501498    NJ   F  1910      Grace     95
3501499    NJ   F  1910       Anne     94
3501500    NJ   F  1910     Evelyn     89
3501501    NJ   F  1910   Gertrude     87
3501502    NJ   F  1910      Edith     86
3501503    NJ   F  1910       Edna     84
3501504    NJ   F  1910    Eleanor     83
3501505    NJ   F  1910      Irene     83
3501506    NJ   F  1910        Ann     82
3501507    NJ   F  1910      Julia     82
3501508    NJ   F  1910     Jennie     76
3501509    NJ   F  1910     Louise     75
...       ...  ..   ...        ...    ...
3651061    NJ   M  2017      Randy      5
3651062    NJ   M  2017     Rayden      5
3651063    NJ   M  2017     Reagan      5
3651064    NJ   M  2017   Reynaldo      5
3651065    NJ   M  2017     Rivaan      5
3651066    NJ   M  2017    Riyansh      5
3651067    NJ   M  2017   Robinson      5
3651068    NJ   M  2017     Rodney      5
3651069    NJ   M  2017       Saad      5
3651070    NJ   M  2017    Santana      5
3651071    NJ   M  2017       Shia      5
3651072    NJ   M  2017   Shivansh      5
3651073    NJ   M  2017    Shreyan      5
3651074    NJ   M  2017       Syed      5
3651075    NJ   M  2017      Sylas      5
3651076    NJ   M  2017     Taylor      5
3651077    NJ   M  2017   Terrance      5
3651078    NJ   M  2017    Terrell      5
3651079    NJ   M  2017      Terry      5
3651080    NJ   M  2017        Tru      5
3651081    NJ   M  2017      Tymir      5
3651082    NJ   M  2017      Tyree      5
3651083    NJ   M  2017    Vaibhav      5
3651084    NJ   M  2017     Vikram      5
3651085    NJ   M  2017     Vishnu      5
3651086    NJ   M  2017     Waylon      5
3651087    NJ   M  2017       Will      5
3651088    NJ   M  2017     Zavier      5
3651089    NJ   M  2017    Zayvion      5
3651090    NJ   M  2017      Zyion      5

[149611 rows x 5 columns]

## Activity 2.5
Select the names in Year 2000 (for all baby_names) that have larger than 3000 counts. What do you notice?

(If you use `p & q` to filter the dataframe, make sure to use `df[df[(p) & (q)]]` or `df.loc[df[(p) & (q)]])`

**Remember** that both slicing and using `loc` will achieve the same result, it is just that `loc` is typically faster in production. You are free to use whichever one you would like.

In [58]:
### BEGIN SOLUTION
result = baby_names.loc[(baby_names['Year']==2000) & (baby_names['Count']>3000)]
### END SOLUTION

In [40]:
result

State Sex  Year         Name  Count
697386     CA   M  2000       Daniel   4339
697387     CA   M  2000      Anthony   3838
697388     CA   M  2000         Jose   3803
697389     CA   M  2000       Andrew   3600
697390     CA   M  2000      Michael   3571
697391     CA   M  2000        Jacob   3520
697392     CA   M  2000       Joshua   3356
697393     CA   M  2000  Christopher   3335
697394     CA   M  2000        David   3280
697395     CA   M  2000      Matthew   3254
5289638    TX   M  2000         Jose   3097

In [41]:
### BEGIN SOLUTION
# complete the assert statements below
assert len(result) == ??
assert result["Count"].sum() == ??
assert result["Count"].iloc[0] == ??
### END SOLUTION


## Activity 2.6 - Data Aggregration (Grouping Data Frames)

To count the number of instances of each unique value in a `Series`, we can use the `value_counts()` [method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) as `df["col_name"].value_counts()`. Count the number of different names for each Year in `NJ` (New Jersey).  (You may use the `nj` DataFrame created above.)

**Note:** *We are not computing the number of babies but instead the number of names (rows in the table) for each year.*

In [60]:
### BEGIN SOLUTION
num_of_names_per_year=nj['Year'].value_counts()
### END SOLUTION

In [61]:
num_of_names_per_year[:5]

2007    2567
2008    2536
2009    2530
2014    2509
2012    2490
Name: Year, dtype: int64

In [44]:
### BEGIN SOLUTION
# complete the assert statements below
assert num_of_names_per_year[2007] == ??
assert num_of_names_per_year[:5].sum() == ??
assert num_of_names_per_year[1910] == ??
assert num_of_names_per_year[:15].sum() == ??
### END SOLUTION

AssertionError: 

## Activity 2.7 - Gender Counts
Count the number of different names for each gender in `NJ`. 

In [71]:
### BEGIN SOLUTION
num_of_names_per_gender = nj['Sex'].value_counts()
### END SOLUTION

In [72]:
num_of_names_per_gender

F    85051
M    64560
Name: Sex, dtype: int64

In [74]:
### BEGIN SOLUTION
# fix the errors in these assertions
assert num_of_names_per_gender['F'] == 100000
assert num_of_names_per_gender['F'] == 100000
assert num_of_names_per_gender['M'] == 100000
### END SOLUTION

AssertionError: 

### Activity 2.7.1

Let's move back to baby names and specifically, the `nj` dataframe. Find the sum of `Count` for each `Name` in the `nj` table. You should use `df.groupby("col_name").sum()`. Your result should be a Pandas Series.

**Note:** *In this question we are now computing the number of registered babies with a given name.*

In [84]:
### BEGIN SOLUTION
nj.groupby("Count").sum()
### END SOLUTION

Year
Count          
5      37646216
6      27157328
7      20857595
8      16594975
9      13686413
10     11692335
11      9597597
12      8669053
13      7460755
14      6646078
15      5858752
16      5272654
17      4744588
18      4480966
19      3912163
20      3772216
21      3359988
22      3216587
23      2929358
24      2828424
25      2611156
26      2328217
27      2322191
28      2243290
29      2009923
30      1955785
31      1839297
32      1764024
33      1731012
34      1615524
...         ...
3209       1961
3218       1958
3246       1946
3270       1965
3280       1949
3281       1964
3307       1952
3333       1959
3337       1948
3353       1946
3357       1960
3366       1953
3375       1961
3398       1962
3418       1954
3427       1955
3434       1951
3443       1956
3450       1963
3463       1957
3467       1956
3468       3910
3474       1957
3475       1955
3479       1961
3480       1953
3533       1954
3565       1947
3671       1947
3700       1964

[1697 rows x 1 columns]

In [85]:
count_for_names.sort_values(ascending=False)[:5]

TypeError: sort_values() missing 1 required positional argument: 'by'

In [50]:
### BEGUN SOLUTION
# fix the assertion errors
assert count_for_names["Michael"] == ??
assert count_for_names[:100].sum() == ??
assert count_for_names["David"] == ??
assert count_for_names[:1000].sum() == ??
### END SOLUTION

SyntaxError: invalid syntax (<ipython-input-50-9a11112ca706>, line 1)

### Activity 2.7.2

Find the sum of `Count` for each female name after year 1999 (`>1999`) in New Jersey.

In [51]:
### BEGIN SOLUTION

### END SOLUTION

In [52]:
female_name_count.sort_values(ascending=False)[:5]

Name
Emily       10164
Isabella    10105
Olivia       9419
Sophia       8765
Emma         8356
Name: Count, dtype: int64

In [53]:
### BEGIN SOLUTION
# fix the assertion errors
assert female_name_count["Emily"] == ??
assert female_name_count[:100].sum() == ??
assert female_name_count["Isabella"] == ??
assert female_name_count[:10000].sum() == ??
### END SOLUTION

SyntaxError: invalid syntax (<ipython-input-53-cbac5c71f684>, line 1)

## Activity 2.8 - Most Popular Names

#### Part 1 What is the most popular female name and male name in any given year for a given state?
Complete a function def most_popular(year, state) that takes a year and a state and return this list.


In [ ]:
### BEGIN SOLUTION
def most_popular(year,state):
   # function_body
### END SOLUTION

In [3]:
### BEGIN SOLUTION
# PART 1 : using the function find the most popular male and female names
# in NJ for year 2000

year = 2000
state = 'NJ'
most_popular(year, state)


### END SOLUTION

NameError: name 'most_popular' is not defined

#### Part 2 What is the most popular female and male name across all years and all states?

In [4]:
### BEGIN SOLUTION


### END SOLUTION

## Activity 2.9 Can you deduce birth sex from the last letter of a person’s name?
This is an open ended activity. We would like you to think about how to answer this question. There can be many ways to do this. 
If you can plot a graph like this, you might be able to answer the question. The x-axis is the probability that a name stars with a letter (y-axis) The plot is done using https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.plot.barh.html

![image.png](attachment:image.png)

In [1]:
# your code to plot the graph as above (you need to consider all states, all years data)
### BEGIN SOLUTION


### END SOLUTION

## Activity 2.10 What is your analysis of the question in Activity 2.9?

#### BEGIN SOLUTION
Please edit this markdown cell
#### END SOLUTION

# Task 3 - Working with Movies
In this example, we will work with the movies.csv file (provided with the zip)

Before we jump into using the `groupby` function in Pandas, let's recap how grouping works in general for tabular data through a guided set of questions based on a small toy dataset of movies and genres. 

**Note:** If you want to see a visual of how grouping of data works, here is a link to an animation: [Groupby Animation](https://rutgers.instructure.com/courses/21255/files?preview=4247256)

**Problem Setting:** The summer 2018, there were a lot of good and bad movies that came out. Below is a dataframe with 5 columns: name of the movie as a `string`, the genre of the movie as a `string`, the first name of the director of the movie as a `string`, the average rating out of 10 on Rotten Tomatoes as an `integer`, and the total gross revenue made by the movie as an `integer`. The point of these guided questions (parts a and b) below is to understand how grouping of data works in general, **not** how grouping works in code. We will worry about how grouping works in Pandas in 7c, which will follow.

Below is the `movies` dataframe we are using, imported from the `movies.csv` file located in the `lab02` directory.

In [86]:
movies = pd.read_csv("movies.csv")
movies

director                      genre  \
0         David         Action & Adventure   
1          Bill                     Comedy   
2           Ron  Science Fiction & Fantasy   
3      Baltasar                      Drama   
4          Bart                      Drama   
5          Gary         Action & Adventure   
6          Drew         Action & Adventure   
7          Brad                  Animation   
8          Jeff                     Comedy   
9          J.A.  Science Fiction & Fantasy   
10      Charles                     Comedy   
11       Gerard                     Horror   
12       Peyton         Action & Adventure   
13       Genndy                  Animation   
14       Rawson         Action & Adventure   
15           Ol                     Comedy   
16  Christopher         Action & Adventure   
17         Marc                     Comedy   

                                    movie  rating    revenue  
0                              Deadpool 2       7  318344544  
1                               Book Club       5   68566296  
2                 Solo: A Star Wars Story       6  213476293  
3                                  Adrift       6   31445012  
4                        American Animals       6    2847319  
5                                Oceans 8       6  138803463  
6                           Hotel Artemis       8    6708147  
7                           Incredibles 2       5  594398019  
8                                     Tag       6   54336863  
9          Jurassic World: Fallen Kingdom       6  411873505  
10                             Uncle Drew       5   42201656  
11                        The First Purge       7   68765655  
12                   Ant-Man and the Wasp       5  208681866  
13  Hotel Transylvania 3: Summer Vacation       5  154418311  
14                             Skyscraper       6   66801215  
15            Mamma Mia! Here We Go Again       8  111705055  
16            Mission: Impossible-Fallout       6  182080372  
17                     Christopher Robbin       6    6786317

## Activity 4.1

If we grouped the `movies` dataframe above by `genre`, how many groups would be in the output and what would be the groups? Assign `num_groups` to the number of groups created and fill in `genre_list` with the names of genres as strings that represent the groups.

In [118]:
### BEGIN SOLUTION
genres=movies.groupby('genre')
num_groups = len(genres)
genre_list = ['Action & Adventure', 'Comedy', 'Science Fiction & Fantasy', 'Drama', 'Animation', 'Horror']
### END SOLUTION

In [119]:
assert num_groups ==6
assert set(genre_list) == set(['Action & Adventure', 'Comedy', 'Science Fiction & Fantasy', 'Drama', 'Animation', 'Horror'])

## Activity 4.2

Whenever we group tabular data, it is usually the case that we need to aggregate values from the ungrouped column(s). If we were to group the `movies` dataframe above by `genre`, which column(s) in the `movies` dataframe would it make sense to aggregate if we were interested in finding how well each genre did in the eyes of people? Fill in `agg_cols` with the column name(s).

In [122]:
### BEGIN SOLUTION
agg_cols = ['rating', 'revenue']
### END SOLUTION

In [123]:
assert set(agg_cols) == set(['rating', 'revenue'])

Now, let's see `groupby` in action, instead of keeping everything abstract. To aggregate data in Pandas, we use the `.groupby()` [function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html). The code below will group the `movies` dataframe by `genre` and find the mean/average revenue and rating for each genre. You can verify you had the same number of groups as what you answered in 7.1. 

In [129]:
# write the code to get the table below
import numpy as np
genres=movies.groupby('genre')
genres.groups
genres['rating'].agg([np.mean,np.average])



mean   average
genre                                        
Action & Adventure         6.333333  6.333333
Animation                  5.000000  5.000000
Comedy                     6.000000  6.000000
Drama                      6.000000  6.000000
Horror                     7.000000  7.000000
Science Fiction & Fantasy  6.000000  6.000000

## Activity 4.3: Grouping Multiple Columns

Let's look at the `movies` dataframe. Which of the following lines of code will output the following dataframe? Write your answer as either 1, 2, 3, or 4. Recall that the arguments to `pd.pivot_table` are as follows: `data` is the input dataframe, `index` includes the values we use as rows, `columns` are the columns of the pivot table, `values` are the values in the pivot table, and `aggfunc` is the aggregation function that we use to aggregate `values`.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rating</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
      <th>8</th>
    </tr>
    <tr>
      <th>genre</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Action &amp; Adventure</th>
      <td>208681866.0</td>
      <td>129228350.0</td>
      <td>318344544.0</td>
      <td>6708147.0</td>
    </tr>
    <tr>
      <th>Animation</th>
      <td>374408165.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Comedy</th>
      <td>55383976.0</td>
      <td>30561590.0</td>
      <td>NaN</td>
      <td>111705055.0</td>
    </tr>
    <tr>
      <th>Drama</th>
      <td>NaN</td>
      <td>17146165.5</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Horror</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>68765655.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Science Fiction &amp; Fantasy</th>
      <td>NaN</td>
      <td>312674899.0</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

1) `pd.pivot_table(data=movies, index='genre', columns='rating', values='revenue', aggfunc=np.mean)`

2) `movies.groupby(['genre', 'rating'])['revenue'].mean()`

3) `pd.pivot_table(data=movies, index='rating', columns='genre', values='revenue', aggfunc=np.mean)`

4) `movies.groupby('revenue')[['genre', 'rating']].mean()`

### Activity 4.4

In [130]:
### BEGIN SOLUTION
answer = movies.groupby('revenue')[['genre', 'rating']].mean()
### END SOLUTION
#4

## Activity 4.5: Merging


Time to put everything together! Merge `movies` and `babynames` to find the number of registered baby names for each director. Only include names that appear in both `movies` and `babynames`.

**Hint:** Take a look at the ``to_frame`` method of a series to do this. 

Your first row should look something like this:

**Note**: It is ok if you have 2 separate columns with names instead of just one column.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>director</th>
      <th>genre</th>
      <th>movie</th>
      <th>rating</th>
      <th>revenue</th>
      <th>Count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>David</td>
      <td>Action &amp; Adventure</td>
      <td>Deadpool 2</td>
      <td>7</td>
      <td>318344544</td>
      <td>371646</td>
    </tr>
  </tbody>
</table>
</table>

In [154]:
### BEGIN SOLUTION
#baby_names.loc[(baby_names['Year']==2000)
#& 
#(baby_names['Count']>3000)]
#
merg=[baby_names,movies]
merged_df = pd.concat(merg)
merged_df.loc[:,:]
### END SOLUTION

C:\Users\avocado\AppData\Local\Programs\Python\Python37-32\lib\site-packages\ipykernel_launcher.py:7: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  import sys


MemoryError: 

In [2]:
### BEGIN SOLUTION
# complete the assertions
assert merged_df.loc[0, 'Count'] == ??
assert merged_df.loc[3, 'Count'] == ?? 
assert merged_df.loc[7, 'Count'] == ??
assert merged_df['Count'].sum() == ??
assert len(merged_df) == ??
### END SOLUTION

SyntaxError: invalid syntax (<ipython-input-2-e86690a5ff58>, line 3)

## Activity 4.6

How many directors in the original `movies` table did not get included in the `merged_df` dataframe? Please explain your answer in 1-2 sentences.

In [ ]:
### BEGIN SOLUTION
answer =
### END SOLUTION

EXPLANATION: 

# Submission Instructions

File Name: Please name the file in the style (section_netID_Lab2.ipynb) 1_adg133_Lab2.ipynb

Submit To: Canvas --> Assignments --> Lab2 

Warning: Failure to follow directions may result in loss of points.

## Feedback  (please edit this cell)
Please provide feedback on this lab. Thank you.
1. How would you rate this lab from a scale of 1-10 (10-best?
2. What components of the lab you enjoyed the most (if any):
3. What parts of the lab, you did not find interesting at all? (if any)
4. What other comments can you provide to improve this lab?

Thanks to many people who proof read this lab and provide good feedback. Also thanks to Berkeley Data Science Group for the suggested ideas for this lab. 